### ISSUE TITLE
Hosting webpage from Flash memory instead of SD card

### ISSUE THREAD

**Issue by karoria** [grblHAL/Plugin_WebUI#1]
This is just a request if we can host http server (index.html.gz) from Flash memory instead of SD card. For Teensy4.1, we have very large amount of flash memory unutilized and I think we should use it unless it has negative implications for hosting webserver. Further, there are chances to accidently delete "www" directory from SD.

**Comment by terjeio** [comment]
Making some _read-only_ files available from flash can be done in the plugin code by converting them to arrays and adding filing system primitives so that the http daemon can read them, I guess that should be made an option. A downside to this is that updates requires a rebuild/reflash.

I believe adding a generic writable flash based fileystem will be a lot more work, so I'll leave that for somebody else to tackle.

**Comment by karoria** [comment]
OK. Got it. But I see rebuilding/reflashing for http server as "positive" point as this is a one time procedure only just like we flash grblHAL, the webui should also part of flashing. This will make webui files read only and hence they will be safe. I like the idea to make it an option in code.

**Comment by karoria** [comment]
Hi @terjeio 
After the recent update regarding littlefs and webui server code, I am still not able to run a webui from ESP3D interface. The basic webpage to upload webui files shows up. I upload index.html.gz from https://github.com/luc-github/ESP3D-WEBUI/blob/3.0/dist/CNC/GRBLHal/index.html.gz. But after upload it does nothing. I also want to know how I can upload my grblTouch webui in littlefs which comprises of loose files index.html, style.css, app.js and other support files. I couldn't find a tool to compress all those files to one index.html.gz. Please help.

**Comment by terjeio** [comment]
>  But after upload it does nothing.

No errors in the browser? Check network traffic by open the developer tools by F12 and reload. No request for the root document? Or waiting forever? Here is what I get in Firefox:

![image](https://user-images.githubusercontent.com/20260062/218697390-40b29321-7d3b-4fc4-b790-7a4224f55bd2.png)

Did you have use the `?forcefallback=yes` argument to open the maintenance page? If so you may have an _index.html_ or _index.html.gz_ file on the SD card that takes precedence.

> I couldn't find a tool to compress all those files to one index.html.gz in case it is compulsory to do so. Please help.

Try with [compressing the index.html](https://duckduckgo.com/?t=ffab&q=windows+create+.gz&ia=web) separately and refer to the other files with absolute paths starting with `/littlefs/`, this might work - but I have not tested it.

Alternatively you have to write your own server.c implementation and [add redirection to littlefs](https://github.com/grblHAL/Plugin_WebUI/blob/d18a9cd7fbea02b22655e37f47e1a5c2905b1a57/server.c#L713-L762) for the files you want to load from there. Note that redirection function will only be called if the file in question is not found on the SD card.

**Comment by karoria** [comment]
> No errors in the browser? Check network traffic by open the developer tools by F12 and reload. No request for the root document? Or waiting forever?

Chrome developer tools shows "Failed to load resource: net::ERR_CONTENT_DECODING_FAILED" after uploading index.html.gz and refreshing the url. Here is screenshot.
![Screenshot 2023-02-14 at 11 32 57 PM](https://user-images.githubusercontent.com/45288223/218820963-f4dbc120-036c-4e0b-8f96-6214ae1f0010.png)

> Did you have use the ?forcefallback=yes argument to open the maintenance page? If so you may have an index.html or index.html.gz file on the SD card that takes precedence.

After using ?forcefallback=yes argument, it defaults to index.html from SD card. I noticed here it doesn't follow path www/index.html instead it opens index.html from root of SD card.

> Try with [compressing the index.html](https://duckduckgo.com/?t=ffab&q=windows+create+.gz&ia=web) separately and refer to the other files with absolute paths starting with /littlefs/, this might work - but I have not tested it. Alternatively you have to write your own server.c implementation and [add redirection to littlefs](https://github.com/grblHAL/Plugin_WebUI/blob/d18a9cd7fbea02b22655e37f47e1a5c2905b1a57/server.c#L713-L762) for the files you want to load from there. Note that redirection function will only be called if the file in question is not found on the SD card.

Ok. Will do as I get time to play with it.

**Comment by terjeio** [comment]
> Chrome developer tools shows "Failed to load resource: net::ERR_CONTENT_DECODING_FAILED"

Then the file has been corrupted somehow - does it work if you place it in the root of the SD card?

> I noticed here it doesn't follow path www/index.html instead it opens index.html from root of SD card.

The search sequnece is the root of the SD card first if the path is relative, if this fails then the redirect function is called.

> But as I copied some files, it shows 100% occupation despite my files were approx only 60KB in total.

How much space the files occupies is related to the sector size in the flash which is 64K for the Teensy 4.1, each file needs a whole number of sectors. In addition sector(s) are needed for the root directory and redundancy/failsafe handling.

A side note: the STM32F7xx driver cannot use littlefs due to the flash sector size beeing 128K and the flash beeing quite small.

> I don't get enough storage space due to this issue.

It can be expanded - I set the size fairly low since the WebUI file is small. It can be [expanded here](https://github.com/grblHAL/iMXRT1062/blob/e22620a280e4cce8e5dc899a53853230bdd8bd35/grblHAL_Teensy4/src/littlefs_hal.c#L35-L43), change FS_SIZE.

**Comment by karoria** [comment]
> Then the file has been corrupted somehow - does it work if you place it in the root of the SD card?
It seems I can't download the file from the said GitHub repo. I have tried safari and chrome browsers but no luck. Will you verify that the index.html.gz file at https://github.com/luc-github/ESP3D-WEBUI/blob/3.0/dist/CNC/GRBLHal/index.html.gz works with webui?

> How much space the files occupies is related to the sector size in the flash which is 64K for the Teensy 4.1, each file needs a whole number of sectors. In addition sector(s) are needed for the root directory and redundancy/failsafe handling.
I got it. This means a single file with say 200kB occupies less space than in my case multiple files with less memory.

> It can be expanded - I set the size fairly low since the WebUI file is small. It can be [expanded here](https://github.com/grblHAL/iMXRT1062/blob/e22620a280e4cce8e5dc899a53853230bdd8bd35/grblHAL_Teensy4/src/littlefs_hal.c#L35-L43), change FS_SIZE.
I changed FS_SIZE to 4096 instead of 512 for my Teensy board and my storage issue is over. I can keep all loose files of grblTouch in uncompressed form and the UI works well. So at present, I am happy without file compression. I have tried 2-3 ways of compression using tar.gz and then renaming it to html.gz but the file integrity is lost and raw data of file is shown instead of web app mostly may be due to index.html doesn't get priority there. So, right now I am leaving it uncompressed as I don't have storage issue for now.

One more point is, I have to type <my IP>/littlefs/index.html to get my UI instead of <my IP> only. What should I change to do it right?

**Comment by terjeio** [comment]
The index.html.gz works for me.

>  I have tried 2-3 ways of compression using tar.gz

Can you attach the compressed file to a comment, you may have to zip it first. BTW - have you tried 7z for compressing to gz?

> I have to type /littlefs/index.html to get my UI instead of only. What should I change to do it right?

Somehow manage to compress index.html to .gz successfully.

> WEBUI_MAX_CLIENTS is set to 4 by default but it still doesn't seem to connect to more than one client at a time.

Only one client can connect at a time _and_ claim control (there is only one set of "serial" buffers). WebUI has a [switch function](https://github.com/grblHAL/Plugin_WebUI/blob/d18a9cd7fbea02b22655e37f47e1a5c2905b1a57/server.c#L110-L170) implemented at websocket level where the latest to take control takes over and the previous is notified about losing it.

If you want to have more than one active client in control you have to figure out how to handle that and add or update code accordingly. It may be possible by adding event handlers to the websocket events, but I am not sure...

**Comment by mvalensky** [comment]
Hello,
I want to use the LITTLEFS without any SD Card definition or pin defined and use this plugin on ESP32. How to set the `my_machine.h`? Are there any further modifications necessary?
If I don't enable the SD Card, but I have compilation errors regarding missing PIN definitions and the in the `server.c`the `sdcard_getfs` is undefined.
Here are my defs:
- #define WEBUI_ENABLE 3 
- #define LITTLEFS_ENABLE 2
- #define TELNET_ENABLE 1
- #define WEBSOCKET_ENABLE 1
- #define FTP_ENABLE 1
- #define HTTP_ENABLE 1

**Comment by terjeio** [comment]
> Are there any further modifications necessary?

Yes, I want to split the SD card plugin in two which would partly allow what you are after. One generic part on top of the virtual filing system (VFS) and one to handle cards. Perhaps the latter bit should be incorporated in the [fs_fatfs](https://github.com/grblHAL/Plugin_SD_card/blob/master/fs_fatfs.c) code. And the WebUI/[core preprocessor code](https://github.com/grblHAL/core/blob/f5aafc1a75e55533a87066038750610715d44db4/driver_opts.h#L515-L520) should be changed to not require the SD card plugin.

**Comment by terjeio** [comment]
FYI WebUI is no longer dependent on the SD card plugin.
